subprojects/messaging/src/main/java/org/gradle/internal/event/DefaultListenerManager.java

```diff
@@ -133,8 +138,8 @@ public ListenerManager createChild() {
         private final ReentrantLock broadcasterLock = new ReentrantLock();
         private ListenerDetails logger;
         private Dispatch<MethodInvocation> parentDispatch;
-        private List<Dispatch<MethodInvocation>> allWithLogger = Collections.emptyList();
-        private List<Dispatch<MethodInvocation>> allWithNoLogger = Collections.emptyList();
+        private ImmutableList<Dispatch<MethodInvocation>> allWithLogger = ImmutableList.of();
+        private ImmutableList<Dispatch<MethodInvocation>> allWithNoLogger = ImmutableList.of();
```

Comment:
This won't work I think. @oehme made the change in the reverse direction a while ago to avoid using Guava in anything that's on the worker classpath: https://github.com/gradle/gradle/commit/8ed96691c0d5ad335b03420cd44485b58400c4b7

Controlling what is and what isn't exposed to workers should be handled via dependencies, of course, but still, we probably don't want to revert his change for now.

Reply:
I also have a hard time imagining how this would make a difference in performance. The classloading cost for Guava on the other hand does make a difference.

Reply:
I removed the Guava usage. I added a check for `isEmpty` to avoid creating iterators for empty lists - IIUC, that showed up in flame graphs.

Reply:
Did it? Low-level stuff like iterators are usually optimized away when the code is hot enough.